### DISCUSSION TITLE
Does this project have APIs so can be used by other languages, like Java?

### DISCUSSION THREAD

**Discussion by nlpsuge** [BurntSushi/ripgrep#2058] [Q&A] [answered]
Hi,

Does this project has APIs so can be used by other Java projects to build some kind of local file search project which runs in a browser?

**Comment by BurntSushi** [comment, marked as answer]
I'm not sure I fully understand your question, but ripgrep does have a `--json` flag that will cause it to emit matches in a structured format that is easier to reading. But it does require that you execute `rg` as a sub-process.

If you mean Rust APIs, ripgrep is split into a bunch of libraries: https://docs.rs/grep. But it's not exactly easy to use at the moment, and there is no library API equivalent of ripgrep itself. You have to put all of the pieces together yourself.

**Reply by nlpsuge** [reply]
Sorry, I have changed the title to 'Does this project have APIs so can be used by other languages, like Java?'.

I have a plan to use ripgrep in a Java project, because I personally use it for years, it's reliable, fast and also fits my need. 

I'm not sure how to use ripgrep in Java language. I did a search and seems that it does has APIs or bindings that other languages like Java can use directly.

I may have to use it via:
1. ProcessBuilder (Like subprocess.Popen(cmd))
2. Java JNA + Rust FFI

But I'm not familiar with Rust.

**Reply by BurntSushi** [reply]
No, there is no C API that ripgrep exposes. In theory, you could go off and write one, but you'd have to learn Rust and it would be _a lot_ of work.

Your best bet is to just run `rg` as a child process and parse the JSON output. That's how VS Code integrates ripgrep to provide its "find in files" features.

**Reply by nlpsuge** [reply]
Thank you very much :)
